AudioSampleBufferCompressor shouldn't buffer 0.1s at all time

jyavenard opened the pull request:
#### de609bdd72dc763504c546ae0e87662ea8d0667f
<pre>
AudioSampleBufferCompressor shouldn&apos;t buffer 0.1s at all time
<a href="https://bugs.webkit.org/show_bug.cgi?id=281958">https://bugs.webkit.org/show_bug.cgi?id=281958</a>
<a href="https://rdar.apple.com/138465744">rdar://138465744</a>

Reviewed by Jer Noble.

Have AudioSampleBufferCompressor always process all available frames and
output them as soon as possible.
When flushing, output all decoded frames, keeping the converter&apos;s internal
buffer so that it can continues where it left off.

It was previously possible to fully drain the encoder by calling finish, but
this would set the AudioConverter to EOF and it would no longer encode samples.
The AudioConverter can differentiate final EOF from just having processed all
content by returning a code other than noErr.

Add a drain() method that will output all data available, a method similar to
finish() but will reset the encoder so it can continue to be used.

Resuming will include samples that will get trimmed from the start upon decoding due to codec delay.

Simplify memory management by only ever using the AudioConverter to encode one packet at a time.
Following this change, the latency of the encoder goes from 0.1s to the duration of an audio frame only.

Covered by existing tests.

* Source/WebCore/PAL/pal/cf/AudioToolboxSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/AudioToolboxSoftLink.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
(WebCore::AudioSampleBufferCompressor::AudioSampleBufferCompressor):
(WebCore::AudioSampleBufferCompressor::drain):
(WebCore::AudioSampleBufferCompressor::flushInternal):
(WebCore::AudioSampleBufferCompressor::initAudioConverterForSourceFormatDescription):
(WebCore::AudioSampleBufferCompressor::sampleBuffer): Rename method to enforce that it can only be used with one packet at a time.
(WebCore::AudioSampleBufferCompressor::provideSourceDataNumOutputPackets):
(WebCore::AudioSampleBufferCompressor::processSampleBuffers):
(WebCore::AudioSampleBufferCompressor::processSampleBuffer):
(WebCore::AudioSampleBufferCompressor::computeBufferSizeForAudioFormat): Deleted.
(WebCore::AudioSampleBufferCompressor::sampleBufferWithNumPackets): Deleted.
(WebCore::AudioSampleBufferCompressor::processSampleBuffersUntilLowWaterTime): Deleted.

Canonical link: <a href="https://commits.webkit.org/285681@main">https://commits.webkit.org/285681@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/501244874226b331680fb2d317f0047d0c5b4942

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73508 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52937 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26316 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/77777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/24746 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75623 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62068 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/722 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/77777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/24746 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76575 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/47852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/63249 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/77777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44465 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/20735 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/23077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/66258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/21084 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79394 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/824 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/965 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63261 "Exiting early after 10 failures. 10 tests run. 1 failures") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/65435 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16171 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9293 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/7472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/787 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/3530 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/817 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/803 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/823 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -94,23 +95,41 @@
+        m_isDraining = false;
+        if (m_converter) {
+            auto error = PAL::AudioConverterReset(m_converter);
+            RELEASE_LOG_ERROR_IF(error, MediaStream, "AudioSampleBufferCompressor AudioConverterReset failed %d", error);
```
Should we reject if reset failed?

jyavenard replied:
the recorder always assume they are resolved; just like flush() and finish() currently.

jyavenard replied:
Im going to make both promises be rejected if an error occurred (for drain and finish)

jyavenard replied:
we don't report debug error in that class anyway, nor to the MediaRecorder.

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -462,48 +478,50 @@
-        if (error) {
+        auto error = AudioConverterFillComplexBuffer(m_converter, audioConverterComplexInputDataProc, this, &numOutputPackets, &fillBufferList, &m_destinationPacketDescriptions);
+
+        bool noMoreData = error == kNoMoreDataErr;
```
hasNoMoreData?

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -94,23 +95,41 @@
     return true;
 }
 
+Ref<GenericPromise> AudioSampleBufferCompressor::drain()
```
Will we have a use for flushInternal(false), or will we use drain() instead?

jyavenard replied:
no we still have a use for `flush()` (`flushInternal(false)`)

a "drain" pulls all samples from the AudioConverter, including its internal buffer, it it no longer usable after that unless reset.

So basically if we are about to record an ongoing audio stream, we'll use flush() and wait if needed for a future audio sample which would output the current one.

A call to drain() is when we're paused, and no more audio is coming. When re-submitting samples to the encoder after a drain, nothing will come out of it for about 50ms (for AAC it generates three frames with the same timestamp that are marked as having skip for 50ms)

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -326,7 +334,7 @@
 
     CMSampleBufferRef rawSampleBuffer;
-    auto error = PAL::CMAudioSampleBufferCreateWithPacketDescriptions(kCFAllocatorDefault, buffer.get(), true, NULL, NULL, m_destinationFormatDescription.get(), numPackets, m_currentNativePresentationTimeStamp, m_destinationPacketDescriptions.data(), &rawSampleBuffer);
+    auto error = PAL::CMAudioSampleBufferCreateWithPacketDescriptions(kCFAllocatorDefault, buffer.get(), true, NULL, NULL, m_destinationFormatDescription.get(), numPackets, m_currentNativePresentationTimeStamp, &m_destinationPacketDescriptions, &rawSampleBuffer);
```
This is only ever true if numPackets is 1.
Given we go in that direction, I would remove numPackets from sampleBufferWithNumPackets and rename it to just sampleBuffer or something like that.

jernoble commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
```diff
@@ -32,7 +32,8 @@
 
 typedef struct opaqueCMSampleBuffer *CMSampleBufferRef;
-typedef struct OpaqueAudioConverter* AudioConverterRef;
+typedef struct OpaqueAudioConverter *AudioConverterRef;
```
This is technically incorrect; AudioConverterRef isn't an Obj-C type, so the indirection symbol goes next to the type.

jyavenard replied:
I thought it was the objC style and that was determined by the file language rather than the object.

jyavenard committed:
AudioSampleBufferCompressor shouldn't buffer 0.1s at all time
https://bugs.webkit.org/show_bug.cgi?id=281958
rdar://138465744

Reviewed by Jer Noble.

Have AudioSampleBufferCompressor always process all available frames and
output them as soon as possible.
When flushing, output all decoded frames, keeping the converter's internal
buffer so that it can continues where it left off.

It was previously possible to fully drain the encoder by calling finish, but
this would set the AudioConverter to EOF and it would no longer encode samples.
The AudioConverter can differentiate final EOF from just having processed all
content by returning a code other than noErr.

Add a drain() method that will output all data available, a method similar to
finish() but will reset the encoder so it can continue to be used.

Resuming will include samples that will get trimmed from the start upon decoding due to codec delay.

Simplify memory management by only ever using the AudioConverter to encode one packet at a time.
Following this change, the latency of the encoder goes from 0.1s to the duration of an audio frame only.

Covered by existing tests.

* Source/WebCore/PAL/pal/cf/AudioToolboxSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/AudioToolboxSoftLink.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
(WebCore::AudioSampleBufferCompressor::AudioSampleBufferCompressor):
(WebCore::AudioSampleBufferCompressor::drain):
(WebCore::AudioSampleBufferCompressor::flushInternal):
(WebCore::AudioSampleBufferCompressor::initAudioConverterForSourceFormatDescription):
(WebCore::AudioSampleBufferCompressor::sampleBuffer): Rename method to enforce that it can only be used with one packet at a time.
(WebCore::AudioSampleBufferCompressor::provideSourceDataNumOutputPackets):
(WebCore::AudioSampleBufferCompressor::processSampleBuffers):
(WebCore::AudioSampleBufferCompressor::processSampleBuffer):
(WebCore::AudioSampleBufferCompressor::computeBufferSizeForAudioFormat): Deleted.
(WebCore::AudioSampleBufferCompressor::sampleBufferWithNumPackets): Deleted.
(WebCore::AudioSampleBufferCompressor::processSampleBuffersUntilLowWaterTime): Deleted.

Canonical link: https://commits.webkit.org/285681@main
